### PR TITLE
productionize `community-ci` workflow

### PR DESCRIPTION
**Pull request by alafanechere**
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7315

This PR introduces a new `community_ci.yml` workflow to run connectors tests and format on fork PR.

## How can this safely work as forks contain untrusted code ⁉️ 
* We use [`pull_request_target` triggers](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target): it means the worfklow logic is the one hosted on the target branch. The fork can't alter the CI behavior. We can't use the `pull_request` event because: it does not provide access to secrets on forks and the workflow logic could be altered by the contribution.
* We use [GitHub environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) with protection rule (`community_ci`): the execution of the test and format jobs is conditioned by a human approval from a pool of selected Airbyters. A human review should identify shady activity like secret extraction or resource exhaustion. If the code is deemed safe the Airbyte reviewer can approve the deployment to the environment, which would run CI jobs.
* Github runners are ephemeral so there's no way of accessing artifacts produced by a previous job. We moreover created dedicated runners for community use cases.

## Changes
* I modified [the `community_ci.yml`](https://github.com/airbytehq/airbyte/pull/37404/files#diff-6abaa453ec6560d6a20e21f05b81a385e5aa4cb77b1a62f433f75def6be328b5) workflow from my spike to clean it up and make it work. **This is the first thing you should review**.
* I made `airbyte-ci` accept a `--git-repo-url` option. This is to enable it to clone forked Airbyte repo to compute the list of modified files.
* I changed the `install-airbyte-ci.yml` action to always install the binary when we're on a fork. This will prevent accidentally installing `airbyte-ci` from source if something shady is introduced to this tool.
* I modified the other existing "non fork" workflow to early exit or skip if they're triggered by a fork. This should guarantee mutual exclusion of what's running in `community_ci.yml` and the other non fork workflows.

### Why did you introduce a new workflow and not change the logic of the existing ones ⁉️ 
I felt that introducing a new workflow dedicated to Community CI was simpler to colocate all the specific (and a bit more complex) logic related to running CI on forks.
If we wanted to modify the other workflows to support running on forks it would require changing their trigger from `pull_request` to `pull_request_target`, which has some drawbacks. It would also introduce a lot more conditional expressions in workflows to check  whether we're on a fork.

We can consider streamlining workflows to work on fork / non fork in the future, but I first want to unblock the community review use case in a simple and atomic manner.

## Demo
Test and format correctly ran in [this workflow execution](https://github.com/airbytehq/airbyte/actions/runs/8798137339/job/24144487326) on [this fork](https://github.com/airbytehq/airbyte/pull/37499).

When a push happens on the PR Github displays this button to start the approval flow:
Screenshot 2024-04-23 at 11 32 47

### CODE REVIEW

**Comment by alafanechere** [comment]
* **#37404** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/37404?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alafanechere and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: .github/workflows/community_ci.yml
Comment (wennergr) [reply .github/workflows/community_ci.yml:103]: What you probably want is:

env:
  head_ref: ${{ github.head_ref }}

File: .github/actions/install-airbyte-ci/action.yml
Comment (natikgadzhi) [inline .github/actions/install-airbyte-ci/action.yml:10]: How much do you think we're saving with binary airbyte-ci overall? What if we just default to `pip install` all the time instead? If this removes 1+ page of yaml code, I'd say the simplification of the CI flow might be worth 0.1 seconds on each run?

File: .github/actions/install-airbyte-ci/action.yml
Comment (natikgadzhi) [inline .github/actions/install-airbyte-ci/action.yml:21]: I have a feeling that this is outdated, but I don't remember. I think the new one is v43?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:80]: This enforces that actions are always from our master branch, and not changed in the fork? Why not just guard the whole run with a check that CI itself isn't changeD?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:96]: Should we skip the token if it speeds things up? Are dagger caches actually getting read from if we run that CI on the _next_, not the same commit?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:100]: Why does format task need GCS creds?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:114]: How are those used? Slightly sus that we need to write something in tests?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:134]: These two steps are used in both commands (format and test) — should we ref them out? Is that a thing in gh actions at all, without making a custom action? ;)

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:166]: Since tests don't write anything to dockerhub, let's strip the credentials and make sure that we do not perform docker login at all.

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:168]: Should we in theory make a separate GSM credential for CI runs? Not sure if this adds any extra security, my logic was that in case something goes wrong, we can rotate one cred, not the main one.

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:148]: Out of curiosity, please tell me that Airbyte CI skips update check when it's running in CI. It skips it, right?

**Review by natikgadzhi — COMMENTED** [review]
I love where this is going. I would love to see creds inputs become optional, dropped from community-ci, and would love to skip some operations that don't give us benefit in community-ci and formatting specifically.

File: .github/actions/install-airbyte-ci/action.yml
Comment (alafanechere) [reply .github/actions/install-airbyte-ci/action.yml:10]: We are saving a significant amount of time by:
* not installing Python
* not downloading deps from pip 

Moreover the Python install github action is rate limited and we often reached the limit last year.

This is why @bnchrch worked on building and releasing an airbyte-ci binary (it also simplifies local install of the tool)

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:80]: Because I think we have no guarantee that the fork is up to date with  main repo's `master` and has the most up to date version of our CI logic in `action`.

I faced this problem during my iteration: I opened a fork, made changes the actions on the main repo, and did not understand why the logic of the local action in `.github` was not updated.

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:96]: > Are dagger caches actually getting read from if we run that CI on the next, not the same commit?

Yes they are. Dagger caches are docker layers and volume cache. For instance:
* If you did not change the dependencies of a connector they will get pulled from the poetry/pip cache volume
* If you modify two connectors on the branch but just made a change to one one your latest commit the cache will be re-used

Comment (alafanechere) [reply .github/workflows/community_ci.yml:96]: We could create a community specific dagger token

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:100]: I can't remember :)

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:114]: `airbyte-ci` pushes per connector status checks to the branch with githup api. So the token we use in this workflow has to have this write permission

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:134]: Yes we can avoid code duplication here. I did not do it yet as I wanted to make sure the full flow works first (#draft-mode) .

File: airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/cli/airbyte_ci.py:148]: [It skips it.](https://github.com/airbytehq/airbyte/blob/0f3e63bd45cf9209a31b96f048503143c3421b9a/.github/actions/run-airbyte-ci/action.yml#L118)

**Comment by alafanechere** [comment]
>  I would love to see creds inputs become optional

@natikgadzhi I agree some action inputs (especially the one with creds) should be cleaned up to be optional.
But I would love to work on this in a follow up effort as it is not preventing the workflow from running ATM.

File: .github/actions/airbyte-ci-requirements/action.yml
Comment (alafanechere) [inline .github/actions/airbyte-ci-requirements/action.yml:1]: This worfklow was not used anywhere

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:71]: Yes it'd indeed be a better practice. Can this be a followup 🙏

**Comment by aaronsteers** [comment]
Regarding security, @natikgadzhi and @alafanechere , are you both confident enough in the security profile here that we can relax the GitHub setting to let workflows run freely on fork PRs? We just relaxed on PyAirbyte because this enables users constant feedback on those non-privileged workflows like **lint**, **format**, **static code tests** and **non-cred-tests** like **unit tests**. 

Before bringing in any `pull_request_target` workflows, we should carefully consider if this limits our options to enable more frequent checks for users. (Today all workflows require approval, even though there is zero security risk. And we're adding a new security vulnerability here that we need to mitigate ourselves.)

**Comment by alafanechere** [comment]
> Regarding security, @natikgadzhi and @alafanechere , are you both confident enough in the security profile here that we can relax the GitHub setting to let workflows run freely on fork PRs? 

No I am not. 
*  A. Anytime there's secret access there's a risk for secret exfiltration as we execute untrusted code.
* B. Anytime we execute untrusted code there's a risk for resource abuse.

*B* can be easily mitigated with carefully picked timeouts and concurrency control.
*A* would require rework of the worfklows an airbyte-ci to not require secrets for non-cred-tests. We currently run one single `airbyte-ci` command which runs all tests, so we pass secrets to it. We could rework the workflow for more granular execution. This is something I described in details in [Phase 2 of my tech spec](https://docs.google.com/document/d/1fY1igh5DAHS9t-AwJxefVHHH98iWbpJjlMxyQz-t0Es/edit#heading=h.hg199cghcqo6)

In any case I believe this should be a follow up effort. We could start by not requiring secrets for format as it's a smaller lift compared to running unit tests automatically.

**Comment by alafanechere** [comment]
> Before bringing in any pull_request_target workflows, we should carefully consider if this limits our options to enable more frequent checks for users.

The `pull_request_target` trigger does not limit the checks frequency. It has the same triggering logic as `pull_requests`. What can throttle frequent checks is the use of Github environment with protection which requires human approval.

File: .github/actions/install-airbyte-ci/action.yml
Comment (natikgadzhi) [reply .github/actions/install-airbyte-ci/action.yml:10]: How does it work without Python tho!? TIL

File: .github/workflows/airbyte-ci-tests.yml
Comment (natikgadzhi) [inline .github/workflows/airbyte-ci-tests.yml:70]: Nit: check if new version of checkout is available. V4 I think?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:37]: I assume this needs to change before we merge?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [reply .github/workflows/community_ci.yml:53]: Ah, so we want to ensure that they ran new versions of tests and such?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:125]: Why change the token here? How is it related? Is this one specific to community ci?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:128]: Can you explain this change a bit? Did we rename the variable before but didn't rename this particular input?

File: .github/workflows/community_ci.yml
Comment (natikgadzhi) [inline .github/workflows/community_ci.yml:82]: Do we remove this in favor of @aaronsteers's approach to formatting, given we'll clean that one up?

**Review by natikgadzhi — APPROVED** [review]
Pre-approving to get this moving, caveated on addressing feedback —I hope to have this in soon so we can then cleanup formatting and get @marcosmarxm to be very happy ;)

File: .github/actions/install-airbyte-ci/action.yml
Comment (alafanechere) [reply .github/actions/install-airbyte-ci/action.yml:10]: We use [PyInstaller](https://pyinstaller.org/en/stable/) to produce a binary from the python package.
We produce this binary on each push to a branch changing airbyte-ci.
The `latest` version is the one produced on push to `master`.
The release logic is maintained in this [workflow](https://github.com/airbytehq/airbyte/blob/master/.github/workflows/airbyte-ci-release.yml#L1)
The command to build the binary is [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pyproject.toml#L57)

File: .github/workflows/community_ci.yml
Comment (alafanechere) [reply .github/workflows/community_ci.yml:128]: I'm just aligning the inputs to what is  currently declared in the `format_check.yml` workflow.

**Pull request merged**
